sql-plugin/src/main/scala/com/nvidia/spark/rapids/conditionalExpressions.scala

```diff
@@ -129,13 +164,34 @@ case class GpuCaseWhen(
     }
   }
 
-  override def columnarEval(batch: ColumnarBatch): Any = {
-    // `elseRet` will be closed in `computeIfElse`.
+  override def columnarEval(batch:ColumnarBatch): Any = {
+    var isAllPredsFalse = true
+    // Check if any predicate is the first all-true, then evaluate its true expression
+    // and return the result.
+    branches.foreach { case (predExpr, trueExpr) =>
+      withResource(GpuExpressionsUtils.columnarEvalToColumn(predExpr, batch)) { p =>
+        if (isAllPredsFalse && isAllTrue(p)) {
+          return GpuExpressionsUtils.columnarEvalToColumn(trueExpr, batch)
+        }
+        isAllPredsFalse = isAllPredsFalse && isAllFalse(p)
+      }
+    }
+
     val elseRet = elseValue
       .map(_.columnarEval(batch))
       .getOrElse(GpuScalar(null, branches.last._2.dataType))
-    branches.foldRight[Any](elseRet) { case ((predicateExpr, trueExpr), falseRet) =>
-      computeIfElse(batch, predicateExpr, trueExpr, falseRet)
+    if (isAllPredsFalse) {
+      // No predicate has a true, so return the else value directly.
+      GpuExpressionsUtils.resolveColumnVector(elseRet, batch.numRows())
+    } else {
+      branches.foldRight[Any](elseRet) { case ((predExpr, trueExpr), falseRet) =>
+        // Here evaluates the predicate expressions again, sacrificing a little of perf
+        // for minimizing the memory size of computations. Because caching the predicates
+        // may get OOM if there are too many branches.
+        withResource(GpuExpressionsUtils.columnarEvalToColumn(predExpr, batch)) { p =>
```

Comment:
Do we really want to double-evaluate the expression?  What if the predicate expression is relatively expensive to evaluate (e.g.: is a regular expression match on a large string column or a UDF)?

Rather than waste work re-evaluating expressions that have unknown time and complexity, I'd rather see an optimization for the common cases where the case/when is really just an if/else or we budget how much memory we're willing to spend on caching the predicate expressions.  In other words, if we see a case/when that has a lot of predicate expressions, maybe we abandon trying to perform the optimization (or maybe just check the very first predicate for the all true case but otherwise fallback) to avoid excessive memory computation.

As it is now, I'm not convinced this is a net-win since we now are making this significantly more expensive in the case where we don't fall entirely in one predicate for each batch.

Reply:
This was for OOM test. Now updated.